webapp/app/controllers/test.js

```diff
@@ -0,0 +1,69 @@
+import Ember from 'ember';
+
+export default Ember.Controller.extend({
+  treeCompliantMetadata: function() {
+    var model = this.get('model');
+
+    var metadata = model.get('testMetadata');
+    var compliantObject = [];
+    this.createSubtree(compliantObject, '', metadata);
+    return compliantObject;
+  }.property('model.testMetadata'),
+
+  createSubtree: function myself(array_to_append, parentIndex, treeObject) {
```

Comment:
No, I think a single component implementing the tree, receiving a nested JSON as a model is a good start. At least that way the complex part will be properly encapsulated...